rpc/grpc/api.go

```diff
@@ -37,3 +49,346 @@ func (bapi *broadcastAPI) BroadcastTx(ctx context.Context, req *RequestBroadcast
 		},
 	}, nil
 }
+
+type BlockAPI struct {
+	sync.Mutex
+	heightListeners      map[chan NewHeightEvent]struct{}
+	newBlockSubscription eventstypes.Subscription
+}
+
+func NewBlockAPI() *BlockAPI {
+	return &BlockAPI{
+		// TODO(rach-id) make 1000 configurable if there is a need for it
+		heightListeners: make(map[chan NewHeightEvent]struct{}, 1000),
+	}
+}
+
+func (blockAPI *BlockAPI) StartNewBlockEventListener(ctx context.Context) {
+	env := core.GetEnvironment()
+	if blockAPI.newBlockSubscription == nil {
+		var err error
+		blockAPI.newBlockSubscription, err = env.EventBus.Subscribe(
+			ctx,
+			fmt.Sprintf("new-block-grpc-subscription-%s", rand.Str(6)),
+			eventstypes.EventQueryNewBlock,
+			500,
+		)
+		if err != nil {
+			env.Logger.Error("Failed to subscribe to new blocks", "err", err)
+			return
+		}
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-blockAPI.newBlockSubscription.Cancelled():
+			env.Logger.Error("cancelled grpc subscription. retrying")
+			ok, err := blockAPI.retryNewBlocksSubscription(ctx)
+			if err != nil {
+				blockAPI.closeAllListeners()
+				return
+			}
+			if !ok {
+				// this will happen when the context is done. we can stop here
+				return
+			}
+		case event, ok := <-blockAPI.newBlockSubscription.Out():
+			if !ok {
+				env.Logger.Error("new blocks subscription closed. re-subscribing")
+				ok, err := blockAPI.retryNewBlocksSubscription(ctx)
+				if err != nil {
+					blockAPI.closeAllListeners()
+					return
+				}
+				if !ok {
+					// this will happen when the context is done. we can stop here
+					return
+				}
+				continue
+			}
+			newBlockEvent, ok := event.Events()[eventstypes.EventTypeKey]
+			if !ok || len(newBlockEvent) == 0 || newBlockEvent[0] != eventstypes.EventNewBlock {
+				continue
+			}
+			data, ok := event.Data().(eventstypes.EventDataNewBlock)
+			if !ok {
+				env.Logger.Debug("couldn't cast event data to new block")
+				continue
```

Comment:
[question] should this log an error and terminate instead of continuing?

```suggestion
				env.Logger.Error("couldn't cast event data to new block")
				return
```

Reply:
I didn't want to have the gRPC logs polluting the output but it's a good point. I will handle this in a separate PR where I will be doing more testing and adding more logic for failures etc. I didn't want to include everything in the same PR to reduce complexity

Reply:
Panicking would mean stopping the node. is this something we want to do? 

That's a separate issue I want to handle in a subsequent PR. If the gRPC connection panics, the node goes down. So I am thinking of having recovers around here or add them to the cosmos-sdk.